frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,907 @@
+	///
+	/// Note: this does not enforce any ordering relation between the submission removed and that
+	/// inserted.
+	///
+	/// Note: this doesn't insert into `insertion_overlay`, the optional new insertion must be
+	/// inserted into  `insertion_overlay` to keep the variable `self` in a valid state.
+	fn swap_out_submission(
+		&mut self,
+		remove_score: ElectionScore,
+		insert: Option<(ElectionScore, u32)>,
+	) -> Option<SignedSubmissionOf<T>> {
+		let remove_idx = self.indices.remove(&remove_score)?;
+		if let Some((insert_score, insert_idx)) = insert {
+			self.indices
+				.try_insert(insert_score, insert_idx)
+				.expect("just removed an item, we must be under capacity; qed");
+		}
+
+		self.insertion_overlay.remove(&remove_idx).or_else(|| {
+			(!self.deletion_overlay.contains(&remove_idx)).then(|| {
+				self.deletion_overlay.insert(remove_idx);
+				SignedSubmissionsMap::<T>::try_get(remove_idx).ok()
+			}).flatten()
+		})
+	}
+
+	/// Iterate through the set of signed submissions in order of increasing score.
+	pub fn iter(&self) -> impl '_ + Iterator<Item = SignedSubmissionOf<T>> {
+		self.indices.iter().filter_map(move |(_score, &idx)| {
+			let maybe_submission = self.get_submission(idx);
+			if maybe_submission.is_none() {
+				log!(
+					error,
+					"SignedSubmissions internal state is invalid (idx {}); \
+					there is a logic error in code handling signed solution submissions",
+					idx,
+				)
+			}
+			maybe_submission
+		})
+	}
+
+	/// Empty the set of signed submissions, returning an iterator of signed submissions in
+	/// arbitrary order.
+	///
+	/// Note that if the iterator is dropped without consuming all elements, not all may be removed
+	/// from the underlying `SignedSubmissionsMap`, putting the storages into an invalid state.
+	///
+	/// Note that, like `put`, this function consumes `Self` and modifies storage.
+	fn drain(mut self) -> impl Iterator<Item = SignedSubmissionOf<T>> {
+		SignedSubmissionIndices::<T>::kill();
+		SignedSubmissionNextIndex::<T>::kill();
+		let insertion_overlay = sp_std::mem::take(&mut self.insertion_overlay);
+		SignedSubmissionsMap::<T>::drain()
+			.filter(move |(k, _v)| !self.deletion_overlay.contains(k))
+			.map(|(_k, v)| v)
+			.chain(insertion_overlay.into_iter().map(|(_k, v)| v))
+	}
+
+	/// Decode the length of the signed submissions without actually reading the entire struct into
+	/// memory.
+	///
+	/// Note that if you hold an instance of `SignedSubmissions`, this function does _not_
+	/// track its current length. This only decodes what is currently stored in memory.
+	pub fn decode_len() -> Option<usize> {
+		SignedSubmissionIndices::<T>::decode_len()
+	}
+
+	/// Insert a new signed submission into the set.
+	///
+	/// The outer `Option` of the return value indicates whether a solution was inserted. In the
+	/// case that nothing was inserted, nothing is ever removed.
+	///
+	/// The inner `Option` of the return value contains the submission which was removed, if any.
```

Comment:
Not sure what rust convention is here; but in the future can we make one of these options an enum with a declaritive name? Or would that make it less ergonomic for users of this method?

Reply:
786898df9 adds that enum. There's a bit of additional `match`-wrangling required, because really we do want `Option<weakest>` much of the time, but it's almost always better to make the situation explicit with a custom enum instead of making users read the docs to figure out what `Option<Option<T>>` means.